Fix ghost visibility

ElectroJr opened the pull request:
## About the PR
Updates ghost system so that ghosts are invisible by default.

## Technical details
Revenants made ghosts default to being visible instead of invisible, while relying on PVS visibility masks to prevent clients from seeing them, but that obviously doesn't work when PVS is disabled and can introduce bugs (e.g., see space-wizards/RobustToolbox/pull/5598).

Ghost sprite visibility and PVS visibility masks are now configurable via `GhostSystem.SetVisible`, and revenants now use that.

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

:cl:
- fix: Fixed ghosts sometimes being visible when they shouldn't be.

keronshb reviewed (changes requested):
Sanity check me on this but I think the Wizard Summon Ghost event needs to be updated to include the new method as well.

keronshb commented:
> > Sanity check me on this but I think the Wizard Summon Ghost event needs to be updated to include the new method as well.
> 
> The action / `ToggleGhostVisibilityToAllEvent` seems to just use `MakeVisible(true)` (same as the end-of-round stuff), so it seems like its behaviour should be unchanged. While I didn't test wizard specifically, the end of round visibility still worked.
> 
> Though there's a separate issue that also exists on master, where `MakeVisible` only affects currently existing ghosts, and won't affect any newly created ones. This probably doesn't matter much for the end-of-round ghost visibility thing, but probably matters somewhat more for wizard. Also the action description is wrong ("invisible" instead of "visible").

The eventual goal is to get it to work like that, but I'm honestly not sure how or where to start for that.
The text is fixed in #34994 (which is in limbo for some reason?)

slarticodefast commented:
@metalgearsloth Is this even needed anymore after your vismask changes?

ElectroJr commented:
> @ metalgearsloth Is this even needed anymore after your vismask changes?

If you mean #35347, that only touched the ghost's eye masks, not their visibility layers.

I've updated the PR to fix conflicts, and also fixed how the show-ghosts wizard action works so that it also applies to newly created ghosts. In the process of fixing conflicts, I've introduced some more breaking changes:
- Renamed `MakeVisible()` to `SetAllObserversVisible()` for clarity.
- Replaced the `AllowGhostShownByEvent` tag introduced in #34994 with a bool datafield on ghost component.
  - Its a tag that exists purely to toggle ghost comp related behaviour, it should've just been a field instead of tag.

ElectroJr commented:
I've re-written most of the PR to remove the revenant code duplication and its more of a refactor than a bugfix now.

MilonPL commented on Content.Client/Ghost/Commands/ToggleGhostVisibilityCommand.cs:
```diff
@@ -12,15 +12,15 @@ public sealed class ToggleGhostVisibilityCommand : IConsoleCommand
     public void Execute(IConsoleShell shell, string argStr, string[] args)
     {
-        var ghostSystem = _entSysMan.GetEntitySystem<GhostSystem>();
+        var ghostSystem = _entSysMan.GetEntitySystem<GhostVisibilitySystem>();
```
Turn it into a LocalizedEntityCommand and use dependencies to get the system?

MilonPL commented on Content.Client/Ghost/GhostVisibilitySystem.cs:
```diff
@@ -0,0 +1,81 @@
+
+    private void OnAttached(Entity<GhostVisibilityComponent> ent, ref PlayerAttachedEvent args)
+    {
+        UpdateVisibility(ent!);
```
This is converting from `Entity<T>` into `Entity<T1?, T2?>` so we can use the `AsNullable` method. There's already implicit conversion happening from `T` to `T1, T2`.

MilonPL commented on Content.Client/Ghost/GhostVisibilitySystem.cs:
```diff
@@ -0,0 +1,81 @@
+
+namespace Content.Client.Ghost;
+
+public sealed class GhostVisibilitySystem: SharedGhostVisibilitySystem
```
`GhostVisibilitySystem : SharedGhostVisibilitySystem`

MilonPL commented on Content.Server/Administration/Commands/ShowGhostsCommand.cs:
```diff
@@ -28,11 +27,7 @@ public void Execute(IConsoleShell shell, string argStr, string[] args)
-
-            ghostSys.MakeVisible(visible);
-            revSys.MakeVisible(visible);
+            _entities.System<SharedGhostVisibilitySystem>().SetAllVisible(visible);
```
Same here I guess, could use dependency injection instead.

MilonPL commented on Content.Shared/Ghost/SharedGhostVisibilitySystem.cs:
```diff
@@ -0,0 +1,137 @@
+            return;
+
+        // VisibilityComponent might not exist yet, and will not get added on client
+        Resolve(ghost.Owner, ref ghost.Comp2, false);
```
What's the point of this? The Resolve above will return if Comp2 is null or can't be found on the entity.

MilonPL commented on Content.Client/Ghost/GhostVisibilitySystem.cs:
```diff
@@ -0,0 +1,81 @@
+
+    private void OnDetached(Entity<GhostVisibilityComponent> ent, ref PlayerDetachedEvent args)
+    {
+        UpdateVisibility(ent!);
```
Same here.
